communication/nats/dialog/dialog.go

```diff
@@ -19,16 +19,21 @@ package dialog
 
 import (
 	"github.com/mysteriumnetwork/node/communication"
+	"github.com/mysteriumnetwork/node/communication/nats/discovery"
 	"github.com/mysteriumnetwork/node/identity"
 )
 
 type dialog struct {
 	communication.Sender
 	communication.Receiver
-	peerID identity.Identity
+	peerID      identity.Identity
+	peerAddress *discovery.AddressNATS
 }
 
 func (dialog *dialog) Close() error {
+	if dialog.peerAddress != nil {
```

Comment:
In which cases peer address can be nil? Those if not nil guards indicate that there is something wrong with struct init - why we cannot have dialog fully constructed and prepared before usage?

Reply:
Usually it is called a defensive programming. Most likely that peerAddress cannot be nil; Not sure if it is compatible with go guidelines, but I used to such coding. Usually e2e tests should expose such nil and it should be fixed in the code that no nil would surface.

If err is produced somewhere in a process of creating such address, then we should terminate early and should not reach Disconnect at all.

Reply:
Defensive programming makes sense in cases where you cannot control your functions input. In here we are constructing private structure with fields. Any unnecessary code lines raise questions

Reply:
The problem here is that we are using this `dialog` struct for both `dialogWaiter` and `dialogEstablisher`. `dialogWaiter` is reusing the single nats connection to all dialogs, but `dialogEstablisher` is creating a new nats connection for every dialog, and this connection should be closed with a dialog. 
I know this is not a perfect architecture, but I don't really want to redo it in the scope of this PR.

Moreover, we are planning to change this communication, so it will be changed. But this PR is a quick fix for a memory leaking.

Reply:
About a defensive part, I prefer to have an extra check, but not an unexpected panic when it will be changed next time.

Reply:
That sounds reasonable enough. I really hope new communication architecture will come soon enough.

Reply:
> About a defensive part, I prefer to have an extra check, but not an unexpected panic when it will be changed next time.

Defensive part is not a single bullet and must be used right, there are some code smells related it. And dont' you think if you find out yourself in dialog with nil that something is terribly wrong anyway long before you reached Close? Do you really want to swallow it and ignore (if not nil does exactly that), fail fast - check in construction that all your params are as expected to be